algorithms/src/snark/marlin/ahp/prover/round_functions/second.rs

```diff
@@ -124,23 +115,137 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
+                        z
+                    })
+                    .sum::<DensePolynomial<F>>();
+                assert!(z.degree() <= constraint_domain.size());
+
+                end_timer!(z_time);
+
+                let circuit_specific_lhs = Self::calculate_circuit_specific_lhs(
+                    &constraint_domain,
+                    fft_precomputation,
+                    ifft_precomputation,
+                    summed_t,
+                    summed_z_m,
+                    z,
+                    *alpha,
+                );
+
+                // Let H = largest_domain;
+                // Let H_i = domain;
+                // Let v_H := H.vanishing_polynomial();
+                // Let v_H_i := H_i.vanishing_polynomial();
+                // Let s_i := H.selector_polynomial(H_i) = (v_H / v_H_i) * (H_i.size() / H.size());
+                // Let m := masking polynomial
+                // Let c_i := circuit combiner
+                // Let lhs_i := circuit_specific partial left hand size of the lincheck sumcheck, see protocol docs for more details
+
+                // Instead of just multiplying each lhs_i by `s_i*c_i`, we reorder the lincheck sumcheck to cancel out some terms
+                // This removes a mul and div by v_H operation over each circuit's (max_constraint_domain - constraint_domain)
+
+                // Later on, we divide each `lhs_i` and m by v_H, then multiply by `s_i*c_i` and sum them.
+                // Substituting in s, we get that:
+                // (\sum_i{lhs_i} + m)/v_H = \sum{h_1_i*v_H + x_g_1_i}
+                // \sum_i{c_i*s_i*(lhs_i/v_H - x_g_1_i)} + (m/v_H - x_g_1_m) = \sum{h_1_i*v_H}
+                // \sum_i{c_i*(H_i.size()/H.size())*(lhs_i/v_H_i - x_g_1_i*v_H/v_H_i)} + (m/v_H - x_g_1_m) = \sum{h_1_i*v_H}
+                // \sum_i{c_i*(H_i.size()/H.size())*(lhs_i/v_H_i} + m/v_H = \sum{h_1_i*v_H} + \sum{c_i*x_g_1_i*(v_H/v_H_i)*(H_i.size()/H.size())} + x_g_1_m
+                // (\sum_i{c_i*s_i*lhs_i} + m)/v_H = \sum{h_1_i*v_H} + \sum{c_i*s_i*x_g_1_i} + x_g_1_m
+                // (\sum_i{c_i*s_i*lhs_i} + m)/v_H = h_1*v_H + x_g_1
+                // That's what we're computing here.
+
+                let sumcheck_time =
+                    start_timer!(|| format!("Compute sumcheck h and g polys for circuit {}", _circuit_id));
+                let (mut h_1_i, mut xg_1_i) = circuit_specific_lhs.divide_by_vanishing_poly(constraint_domain).unwrap();
+                xg_1_i = xg_1_i.mul_by_vanishing_poly(max_constraint_domain);
+                let (mut xg_1_i, remainder) = xg_1_i.divide_by_vanishing_poly(constraint_domain).unwrap();
+                assert!(remainder.is_zero());
+                let multiplier =
+                    circuit_combiner * constraint_domain.size_as_field_element * max_constraint_domain_inverse;
+                cfg_iter_mut!(h_1_i.coeffs).for_each(|c| *c *= multiplier);
+                cfg_iter_mut!(xg_1_i.coeffs).for_each(|c| *c *= multiplier);
```

Comment:
I think we can multiply circuit_specific LHS by `circuit_combiner` once before the division, instead of  multiplying both `h_1` and `xg_1` separately by `circuit_combiner`.

Also IMO a simpler way to phrase this code would be to multiply each `circuit_specific_lhs` by `selector * circuit_combiner`, sum up these `lhs`-es to get a cumulative LHS, add in the mask_poly to this cumulative LHS, and then finally divide the latter by `max_constraint_vanishing_poly` once to get the actual h_1 and xg_1.

Reply:
> I think we can multiply circuit_specific LHS by circuit_combiner once before the division, instead of multiplying both h_1 and xg_1 separately by circuit_combiner.

You are right, adjusted!

> Also IMO a simpler way to phrase this code would be to multiply each circuit_specific_lhs by selector * circuit_combiner, sum up these lhs-es to get a cumulative LHS, add in the mask_poly to this cumulative LHS, and then finally divide the latter by max_constraint_vanishing_poly once to get the actual h_1 and xg_1.

This sounds like the original setup: ea85c95ea
Which indeed is simpler, but doesn't take advantage of cancelling out terms? And if we do want to cancel out terms, we're stuck with this approach I think, as outlined in the big comment of that commit

Reply:
Hm I was thinking that switching back to the old way might save memory, but it perhaps won't.